tests: avoid t.Parallel() when the top test has defers

knz opened the pull request:
Sub-tests that invoke `t.Parallel()` get to run concurrently with
their parent test, and may be delayed arbitrarily past beyond the end
of the termination of the parent test (including beyond the execution
of its `defer` calls).

This means it's unsafe to call `t.Parallel()` with
e.g. `leaktest.AfterTest()`.

Release note: None

knz committed:
tests: avoid t.Parallel() when the top test has defers

Sub-tests that invoke `t.Parallel()` get to run concurrently with
their parent test, and may be delayed arbitrarily past beyond the end
of the termination of the parent test (including beyond the execution
of its `defer` calls).

This means it's unsafe to call `t.Parallel()` with
e.g. `leaktest.AfterTest()`.

This patch removes `t.Parallel` from the test in
`sql/physical_props_test.go` and modifies
`pkg/rpc/TestGRPCKeepaliveFailureFailsInflightRPCs` to use a wait
group instead.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37072)
<!-- Reviewable:end -->

tbg reviewed (commented):
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @RaduBerinde)

___
*[pkg/rpc/context_test.go, line 803 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDsmt821zU9tkMzw9J:-LdDsmt821zU9tkMzw9K:b-7q9upq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/rpc/context_test.go#L803)):*
> ```Go
> 		pName := fmt.Sprintf("client%sserver,server%sclient", connIcon(c.partitionC2S), connIcon(c.partitionS2C))
> 		t.Run(kaName+"/"+pName, func(t *testing.T) {
> 			// TODO(knz): t.Parallel disabled until leaktest.AfterTest learns
> ```

Hmm, that's a bummer. Test takes 28s (up from <4s with enough parallelism). I think here I'd rip out the subtests and start them as goroutines with a waitgroup, writing into an error channel. That way we get even better parallelism than with `t.Parallel` and the main test will wait. Or, if we decide that sometimes use of `t.Parallel` is ok and we don't want to lint it away, we can keep using it and we'll just add the waitgroup to the test.

```go

var wg sync.WaitGroup
wg.Add(len(testCases))
errCh := make(chan error, len(testCases))
for _, c := range testCases {
  go func(c testCase) {
    errCh <- errors.Wrapf(runTestCase(c), "%+v", c)
  }(c)
}
wg.Wait()
close(errCh)

for err := range errCh {
  t.Errorf("%+v", err)
}
```

___
*[pkg/sql/physical_props_test.go, line 363 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDtu3g-hBRcDNo7wxT:-LdDtu3g-hBRcDNo7wxU:b-lldwgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/sql/physical_props_test.go#L363)):*
> ```Go
> 				// how to wait for all sub-tests to complete.
> 				//
> 				// t.Parallel()
> ```

This one doesn't change the duration of the test (I verified), so I'd just nuke this.
___
*[pkg/sql/physical_props_test.go, line 572 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDtxwR1pLX7S2Z5Hg3:-LdDtxwR1pLX7S2Z5Hg4:bt5709s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/sql/physical_props_test.go#L572)):*
> ```Go
> 			// how to wait for all sub-tests to complete.
> 			//
> 			//			t.Parallel()
> ```

Verified that this can be nuked as well.
___
*[pkg/sql/physical_props_test.go, line 769 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDuANhCc6vsTzHpxlg:-LdDuANhCc6vsTzHpxlh:b2wm772) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/sql/physical_props_test.go#L769)):*
> ```Go
> 		tc := testCases[i]
> 		t.Run(tc.name, func(t *testing.T) {
> 			// TODO(knz): t.Parallel disabled until leaktest.AfterTest learns
> ```

Not tested, but pretty sure all the tests in this file don't need `t.Parallel`.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
Interesting. Seems like there is a missing API in `test` that we should file an issue for - either a function to register a cleanup to happen after the parent waits for child subtests, or a function to wait for all subtests.

We could use a wait group though, replacing every `t.Parallel()` call with the block `wg.Add(1); defer wg.Done(); t.Parallel()`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-:-LdDuwvP8PWtvoWadqDy:bw6685v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[pkg/rpc/context_test.go, line 803 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDsmt821zU9tkMzw9J:-LdDvd0g60SxgPFlQBTv:b-jqds0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/rpc/context_test.go#L803)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Hmm, that's a bummer. Test takes 28s (up from <4s with enough parallelism). I think here I'd rip out the subtests and start them as goroutines with a waitgroup, writing into an error channel. That way we get even better parallelism than with `t.Parallel` and the main test will wait. Or, if we decide that sometimes use of `t.Parallel` is ok and we don't want to lint it away, we can keep using it and we'll just add the waitgroup to the test.

```go

var wg sync.WaitGroup
wg.Add(len(testCases))
errCh := make(chan error, len(testCases))
for _, c := range testCases {
  go func(c testCase) {
    errCh <- errors.Wrapf(runTestCase(c), "%+v", c)
  }(c)
}
wg.Wait()
close(errCh)

for err := range errCh {
  t.Errorf("%+v", err)
}
</blockquote></details>

Ah, I made the same suggestion. Why do we need the error channel though? The test infrastructure should still work, we just need the `Wait()`.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @RaduBerinde)

___
*[pkg/rpc/context_test.go, line 803 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDsmt821zU9tkMzw9J:-LdDwFbE0illyZ4VXEUo:b-96dzg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/rpc/context_test.go#L803)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah, I made the same suggestion. Why do we need the error channel though? The test infrastructure should still work, we just need the `Wait()`.
</blockquote></details>

Using `testing.T` on a goroutine is not supported by the testing framework, and will cause problems. In particular, `t.Fatal` off the test goroutine will result in an uncaught `runtime.Goexit`. Note how `runTestCase` doesn't even get to use the `t`. That's to avoid this problem in the first place. Using `t.Error` is fine btw, but that's not a pattern I want to put out there.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/rpc/context_test.go, line 803 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDsmt821zU9tkMzw9J:-LdDziVC64DnDZsC-ZTh:bah06a8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/rpc/context_test.go#L803)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Using `testing.T` on a goroutine is not supported by the testing framework, and will cause problems. In particular, `t.Fatal` off the test goroutine will result in an uncaught `runtime.Goexit`. Note how `runTestCase` doesn't even get to use the `t`. That's to avoid this problem in the first place. Using `t.Error` is fine btw, but that's not a pattern I want to put out there.
</blockquote></details>

But you don't need a goroutine. Just call `Run` like before and put `wg.Done()` inside that. `Run` will return when the inner test calls `Parallel` (it has to, or the parent can't continue to call more `Run`s).


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Raphael points out that the subtests block inside `Parallel()` until the parent test completes, so my suggestion won't work. I filed https://github.com/golang/go/issues/31651.

knz reviewed (commented):
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-:-LdE77cK02x7rF3Ylqfp:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/rpc/context_test.go, line 803 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDsmt821zU9tkMzw9J:-LdE0nil1CZVEA76DW2u:bjh5dap) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/rpc/context_test.go#L803)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

But you don't need a goroutine. Just call `Run` like before and put `wg.Done()` inside that. `Run` will return when the inner test calls `Parallel` (it has to, or the parent can't continue to call more `Run`s).
</blockquote></details>

Modified to use Tobias' suggestion.
___
*[pkg/sql/physical_props_test.go, line 363 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDtu3g-hBRcDNo7wxT:-LdE6ojNEEhX2ddDSW0r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/sql/physical_props_test.go#L363)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This one doesn't change the duration of the test (I verified), so I'd just nuke this.
</blockquote></details>

Done.
___
*[pkg/sql/physical_props_test.go, line 572 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDtxwR1pLX7S2Z5Hg3:-LdE6pAw3vx7f-j3Cbp-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/sql/physical_props_test.go#L572)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Verified that this can be nuked as well.
</blockquote></details>

Done.
___
*[pkg/sql/physical_props_test.go, line 769 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDuANhCc6vsTzHpxlg:-LdE6pg58CBqo6A52yC7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/sql/physical_props_test.go#L769)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Not tested, but pretty sure all the tests in this file don't need `t.Parallel`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

optional suggestion: add a lint against t.Parallel following the example here (ironic that there would be a call to `t.Parallel()` in that same snippet):

https://github.com/cockroachdb/cockroach/blob/e3cbf3354c13f2bd061bc4eaec8ae80854fbecf2/pkg/testutils/lint/lint_test.go#L715-L726

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-:-LdE9ZuW1yJAAqfj8VhX:b566bht)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @RaduBerinde)

___
*[pkg/rpc/context_test.go, line 803 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDsmt821zU9tkMzw9J:-LdE93eo1Ovn0JC97Cr6:b-5px1y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/rpc/context_test.go#L803)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Modified to use Tobias' suggestion.
</blockquote></details>

Optional nit: move this out of the test so that it doesn't have access to a `*testing.T`, or do `t := (*testing.T)(nil)` here to reliably catch any accidental calls in the method.


<!-- Sent from Reviewable.io -->

knz committed:
pkg/rpc: extract the test code from TestGRPCKeepaliveFailureFailsInflightRPCs

Since this now runs on its own goroutine, it is invalid for it to
access the surrounding `testing.T`. This patch moves the code so that
`t` is clearly out of scope.

Release note: None

knz committed:
lint: add a linter to forbid uses of t.Parallel()

This suggests using `sync.WaitGroup` instead.

Release note: None

knz reviewed (commented):
> optional suggestion: add a lint against t.Parallel 

Good idea. Done. Will merge this PR once #36952 has merged, as this will remove the one remaining call to t.Parallel().

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-:-LdESLr0DIcPFAGdBtde:bonwrr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg)

___
*[pkg/rpc/context_test.go, line 803 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDsmt821zU9tkMzw9J:-LdEQCFrELs2G5AunqIf:bv07rih) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/rpc/context_test.go#L803)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Optional nit: move this out of the test so that it doesn't have access to a `*testing.T`, or do `t := (*testing.T)(nil)` here to reliably catch any accidental calls in the method.
</blockquote></details>

Done. This was a good call -- there was a remaining use of `t` inside the function.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-:-LdEWYfH2eOdjOYpHfRr:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/rpc/context_test.go, line 803 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdDsmt821zU9tkMzw9J:-LdEWFdW6AXUJCrKH3p6:baxrgpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce21a47761b5f92a934750445166c3f4c0c99ca1/pkg/rpc/context_test.go#L803)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done. This was a good call -- there was a remaining use of `t` inside the function.
</blockquote></details>

:bug: :hammer: 
___
*[pkg/testutils/lint/lint_test.go, line 775 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdEWTBG8sZT88IiyAYS:-LdEWTBG8sZT88IiyAYT:bhv2v0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb5f1d6be4760940237a2fc0ea0ecf372a770c5e/pkg/testutils/lint/lint_test.go#L775)):*
> ```Go
> 			stream.GrepNot(`// SAFE FOR TESTING`),
> 		), func(s string) {
> 			t.Errorf("\n%s <- forbidden, use a sync.WaitGroup instead", s)
> ```

Refer to https://github.com/golang/go/issues/31651 in the message to satisfy the curious.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/testutils/lint/lint_test.go, line 775 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37072#-LdEWTBG8sZT88IiyAYS:-LdEXQHH1LLMrNHt_zhJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb5f1d6be4760940237a2fc0ea0ecf372a770c5e/pkg/testutils/lint/lint_test.go#L775)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Refer to https://github.com/golang/go/issues/31651 in the message to satisfy the curious.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37072)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz commented:
bors r=tbg,RaduBerinde
